File: 2023/nim/p04.nim
import strutils, sequtils, re, math

let input = readFile("../data/04.txt").strip.splitLines
var cards = repeat(1, input.len)
var score = 0

for i, line in input:
    let
        parts = line.split(re"[:\|]")
        a = parts[1].strip.splitWhitespace.mapIt(it.strip.parseInt)
        b = parts[2].strip.splitWhitespace.mapIt(it.strip.parseInt)
        c = b.countIt(a.contains(it))
    score += 2^c div 2
    for j in (i+1..i+c):
        cards[j] += cards[i]

# part 1
echo score

# part 2
echo cards.sum

File: 2023/ruby/p04.rb
input = IO.foreach('../data/04.txt').map(&:strip)
cards = [1]*input.size
score = input.each_with_index.map{|line,i|
  parts = line.split(/[:|]/)
  a = parts[1].split.map(&:to_i)
  b = parts[2].split.map(&:to_i)
  c = b.count{|n| a.include?(n)}
  (i+1).upto(i+c){|j| cards[j] += cards[i] }
  (2**c)/2
}.sum

puts score, cards.sum
